src/simulators/statevector/qubitvector.hpp

```diff
@@ -772,20 +772,27 @@ void QubitVector<data_t>::zero() {
 
 template <typename data_t>
 void QubitVector<data_t>::set_num_qubits(size_t num_qubits) {
+
+  size_t prev_num_qubits = num_qubits_;
   num_qubits_ = num_qubits;
   data_size_ = BITS[num_qubits];
 
-  // Free any currently assigned memory
-  if (data_)
-    free(data_);
-
   if (checkpoint_) {
     free(checkpoint_);
     checkpoint_ = nullptr;
   }
 
+  // Free any currently assigned memory
+  if (data_) {
+    if (prev_num_qubits != num_qubits_) {
+      free(data_);
+      data_ = nullptr;
+    }
+  }
+
   // Allocate memory for new vector
-  data_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * data_size_));
+  if (data_ == nullptr)
```

Comment:
What if `data_` is not `nullptr`? Shall we reallocate memory to fit with `data_size_`? Are we supposing that `data_size_` hasn't changed so there's no real need for reallocation?

Reply:
`AER::Base::State` keeps `state_t qreg_`, which is a `QV::QubitVector`. `State` is reused in `run_circuit_with_noise` and `run_circuit_without_noise` in `AER::Simulator::QasmController` across each shot simulation.